src/client_side.cc

```diff
@@ -3904,63 +3903,59 @@ ConnStateData::unpinConnection(const bool andClose)
-        // error details, but that context may leave unparsed bytes behind.
-        // Consume them to stop checkLogging() from logging them again later.
-        const auto intputToConsume =
-#if USE_OPENSSL
-            parsingTlsHandshake ? "TLS handshake" : // more specific than CONNECT
-#endif
-            bodyPipe ? "HTTP request body" :
-            pipeline.back()->mayUseConnection() ? "HTTP CONNECT" :
-            nullptr;
-
-        while (const auto context = pipeline.front()) {
-            context->noteIoError(error, lte);
-            context->finished(); // cleanup and self-deregister
-            assert(context != pipeline.front());
-        }
+    const auto leftovers = handleLeftovers();
+
+    while (const auto context = pipeline.front()) {
+        context->noteIoError(error, lte);
+        context->finished(); // cleanup and self-deregister
+        assert(context != pipeline.front());
+    }
 
-        if (intputToConsume && !inBuf.isEmpty()) {
-            debugs(83, 5, "forgetting client " << intputToConsume << " bytes: " << inBuf.length());
-            inBuf.clear();
+    if (!terminatedAll) {
+        if (leftovers || !pipeline.nrequests) {
+            /* Create a temporary ClientHttpRequest object. Its destructor will log. */
+            ClientHttpRequest http(this);
+            http.req_sz = inBuf.length();
+            // XXX: Or we died while waiting for the pinned connection to become idle.
+            http.setErrorUri("error:transaction-end-before-headers");
+            http.updateError(error);
         }
     }
 
+    terminatedAll = true;
+
     clientConnection->close();
 }
 
-/// log the last (attempt at) transaction if nobody else did
-void
-ConnStateData::checkLogging()
+/// whether we should ignore unparsed inBuf bytes
+/// during transaction termination
+bool
+ConnStateData::handleLeftovers()
 {
-    // to simplify our logic, we assume that terminateAll() has been called
-    assert(pipeline.empty());
+    const auto intputToConsume =
+#if USE_OPENSSL
+        parsingTlsHandshake ? "TLS handshake" : // more specific than CONNECT
+#endif
+        bodyPipe ? "HTTP request body" :
+        (!pipeline.empty() && pipeline.back()->mayUseConnection()) ? "HTTP CONNECT" :
+        nullptr;
 
-    // do not log connections that closed after a transaction (it is normal)
-    // TODO: access_log needs ACLs to match received-no-bytes connections
-    if (pipeline.nrequests && inBuf.isEmpty())
-        return;
+    if (intputToConsume && !inBuf.isEmpty()) {
+        debugs(83, 5, "forgetting client " << intputToConsume << " bytes: " << inBuf.length());
+        inBuf.clear();
+    }
 
-    /* Create a temporary ClientHttpRequest object. Its destructor will log. */
-    ClientHttpRequest http(this);
-    http.req_sz = inBuf.length();
-    // XXX: Or we died while waiting for the pinned connection to become idle.
-    http.setErrorUri("error:transaction-end-before-headers");
-    http.updateError(bareError);
```

Comment:
It looks like bareError became used after removing checkLogging() - we can probably remove it too.